python/mlcroissant/mlcroissant/_src/tests/graphs/1.0/distribution_bad_type/output.txt

```diff
@@ -1,2 +1,4 @@
 Found the following 1 error(s) during the validation:
-  -  "a-csv-table" should have an attribute "@type": "http://mlcommons.org/croissant/FileObject" or "@type": "http://mlcommons.org/croissant/FileSet". Got https://schema.org/WRONG_TYPE_HERE instead.
\ No newline at end of file
+  -  "a-csv-table" should have an attribute "@type": "http://mlcommons.org/croissant/FileObject" or "@type": "http://mlcommons.org/croissant/FileSet". Got https://schema.org/WRONG_TYPE_HERE instead.
+Found the following 1 warning(s) during the validation:
+  -  `distribution` is empty for this dataset. If this is not expected, this might be because of bad `@id`s in the jsonld. Note that valid `@id`s cannot contain any whitespace characters to be successfully parsed by the rdflib.
```

Comment:
The distribution has both a bad type and a bad `@id`?

Reply:
No, the distribution is empty for the dataset, because it has a wrong type. The error just states that it _might_ be because of bad `@id` in the distribution (but not only, as in this case).

Reply:
@marcenacp Just passing through to mention that this sort of dovetails with the conversation re: encoding. Rather than worry about what `rdflib` can/does support, it might be clearer to communicate that `@id`s are just URIs, and thus can have whitespace if URL encoded as `%20`.

Reply:
Agreed that we do not need to cite the rdflib -- just committed a newer version of the code.

However, from a few tests I run myself, an `@id` such as `data%20frame` raises a `GenerationError`: it seems to be caused by `rdflib.graph.parse()` in [mlcroissant/_src/core/json_ld.py](https://github.com/mlcommons/croissant/blob/main/python/mlcroissant/mlcroissant/_src/core/json_ld.py) (it will expand it as if it would be a blank node identifier starting with `_` (e.g. `_:Na83c277b00434aa990d19a078d3ae785`) as it does when no `@id` is given, such in the case of `Metadata` or `Dataset`.

So URL-escaping the spaces doesn't seem to be a workaround, we actually can't have them, unless there is a fix (e.g. some parameter that we can tune in the `rdflib`, but from a quick search I couldn't find any obvious way to do so).

From what I see, however, spaces seem to be the only reserved characters  that cause problems out of all [URIs reserved characters](https://documentation.mapp.com/1.0/en/url-encoding-and-what-characters-are-valid-in-a-uri-36147771.html#id-.URLEncodingandwhatcharactersarevalidinaURIv1.0-EncodinganemailaddressinaURI).

Reply:
I'm just a passerby, so don't let my comments hold up any progress here. I'm making a mental note to investigate this on my end. For context, at Kaggle, we rely as much as possible on what users provided to avoid baking too much "magic" into how we represent Croissant from their data. As a result, `@id` is built from the filename, which can have whitespace (among other characters). Ideally, we can just use those as-is, though I understand that if dependency issues are getting in the way, it may be unavoidable.

Bigger picture, this seems like kind of a rough thing for our users to have to be aware of about our tooling (which departs from the `@id` spec from JSON-LD itself as simply URIs which should support URL encoding). Again, not something to hold up any particular merge (including this one), but more to say that I think we may want to go after it sooner rather than later. To that point, I will be tinkering on my end with a Kaggle dataset that has whitespace in the file name (and `@id`) to see what options we may have. Hopefully there's a way but it's just not well-documented? 🙏 🤞